lib/symboldatabase.cpp

```diff
@@ -1576,16 +1576,66 @@ static std::string getIncompleteNameID(const Token* tok)
     return result + tok->expressionString();
 }
 
+namespace {
+    struct ExprIdKey {
+        std::string parentOp;
+        nonneg int operand1;
+        nonneg int operand2;
+        bool operator<(const ExprIdKey& k) const {
+            if (operand1 != k.operand1)
+                return operand1 < k.operand1;
+            if (operand2 != k.operand2)
+                return operand2 < k.operand2;
+            return parentOp < k.parentOp;
+        }
+    };
+    using ExprIdMap = std::map<ExprIdKey, nonneg int>;
+    void setParentExprId(Token* tok, ExprIdMap& exprIdMap, nonneg int &id) {
+        if (!tok->astParent() || tok->astParent()->isControlFlowKeyword())
+            return;
+        const Token* op1 = tok->astParent()->astOperand1();
+        if (op1 && op1->exprId() == 0)
+            return;
+        const Token* op2 = tok->astParent()->astOperand2();
+        if (op2 && op2->exprId() == 0)
+            return;
+
+        if (tok->astParent()->isExpandedMacro()) {
+            tok->astParent()->exprId(id);
+            ++id;
+            setParentExprId(tok->astParent(), exprIdMap, id);
+            return;
+        }
+
+        ExprIdKey key;
+        key.parentOp = tok->astParent()->str();
+        key.operand1 = op1 ? op1->exprId() : 0;
+        key.operand2 = op2 ? op2->exprId() : 0;
+
+        if (op1 && op2 && key.parentOp == "+" && key.operand1 > key.operand2)
+            std::swap(key.operand1, key.operand2);
+
+        const auto it = exprIdMap.find(key);
+        if (it == exprIdMap.end()) {
+            exprIdMap[key] = id;
+            tok->astParent()->exprId(id);
+            ++id;
+        } else {
+            tok->astParent()->exprId(it->second);
+        }
+        setParentExprId(tok->astParent(), exprIdMap, id);
+    }
+}
+
 void SymbolDatabase::createSymbolDatabaseExprIds()
 {
-    nonneg int base = 0;
     // Find highest varId
-    for (const Variable *var : mVariableList) {
-        if (!var)
-            continue;
-        base = std::max<MathLib::bigint>(base, var->declarationId());
+    nonneg int maximumVarId = 0;
+    for (const Token* tok = mTokenizer.list.front(); tok; tok = tok->next()) {
+        if (tok->varId() > maximumVarId)
+            maximumVarId = tok->varId();
```

Comment:
This seems like it would be slower than iterating over the variables. Is this because we set `varId` for variables that dont have a `Variable` class?

Reply:
>  Is this because we set varId for variables that dont have a Variable class?

yes.